File: src/core/context_controller.ts
/**
 * context_controller.ts — Unified Authorization Engine
 *
 * All sensitive operations must pass through this controller's
 * 5-level priority decision chain. Uses fail-closed default deny.
 *
 * Ported from ZCode loop_core/context_controller.py.
 */

import { readFileSync, existsSync } from "node:fs";
import { join, resolve, relative, normalize } from "node:path";
import { parseDocument } from "yaml";
import { loadState, loadGates } from "./state-machine.js";
import type { ProjectState, GatesRegistry, GateDefinition } from "../types/index.js";

// ── Enums ────────────────────────────────────────────────────────────────────

/** Action types that require authorization */
export enum Action {
  WRITE_FILE = "WRITE_FILE",
  EDIT_FILE = "EDIT_FILE",
  EXEC_BASH = "EXEC_BASH",
  LAUNCH_ROLE = "LAUNCH_ROLE",
  INSTALL_PACKAGE = "INSTALL_PACKAGE",
  DELETE_FILE = "DELETE_FILE",
  MODIFY_GATE = "MODIFY_GATE",
  MODIFY_STATE = "MODIFY_STATE",
  SUBMIT_EVIDENCE = "SUBMIT_EVIDENCE",
  CREATE_HANDOFF = "CREATE_HANDOFF",
  ADVANCE_PHASE = "ADVANCE_PHASE",
}

/** Authorization decision */
export enum Decision {
  ALLOW = "ALLOW",
  DENY = "DENY",
  ASK_USER = "ASK_USER",
}

// ── Interfaces ───────────────────────────────────────────────────────────────

/** Authorization request */
export interface AuthRequest {
  action: Action;
  target_path?: string;
  task_id?: string;
  role_id?: string;
  command?: string;
  project_root: string;
}

/** Authorization result */
export interface AuthResult {
  decision: Decision;
  reason: string;
  required_gate_id?: string;
  allowed_paths?: string[];
  /** Which check level produced this result */
  check_level: number;
}

// ── Constants ────────────────────────────────────────────────────────────────

/** Protected paths that always require user approval */
export const PROTECTED_PATHS: string[] = [
  "AGENTS.md",
  ".ai/state.yaml",
  ".ai/gates.yaml",
];

/** Dangerous command patterns that are always denied for EXEC_BASH */
const DANGEROUS_COMMANDS: RegExp[] = [
  /\brm\s+(-[a-zA-Z]*[rf][a-zA-Z]*\s+)*\//,       // rm -rf / or rm file
  /\bformat\s+[a-zA-Z]:/,                           // Windows format
  /\bdd\s+.*\bof=\/dev\/[sh]d/,                     // dd to disk
  /\bmkfs\b/,                                        // mkfs
  /\b(del|remove)\s+\/[sSfq]/,                       // Windows del /S /F /Q
  /\bshutdown\b/,                                    // shutdown
  /\breboot\b/,                                      // reboot
  /\breg\s+delete\b/,                                // Windows registry delete
];

/** Governance file directory pattern */
const GOVERNANCE_DIR = ".ai";

/** Task graph file path relative to project root */
const TASK_GRAPH_FILE = join(".ai", "task_graph.yaml");

// ── Task types (local) ───────────────────────────────────────────────────────

/** Minimal task record shape for task scope checking */
interface TaskRecord {
  task_id: string;
  status: string;
  allowed_paths?: string[];
}

/** Minimal task graph shape */
interface TaskGraph {
  tasks: TaskRecord[];
}

// ── Helper Functions ─────────────────────────────────────────────────────────

/**
 * Check if a file path points to a governance file under .ai/.
 *
 * @param filePath - The file path to check (absolute or relative)
 * @returns `true` if the path is a governance file
 */
export function isGovernancePath(filePath: string): boolean {
  const normalized = normalize(filePath).replace(/\\/g, "/");
  // Match .ai/ prefix or /.ai/ in the path
  return (
    normalized.startsWith(".ai/") ||
    normalized.includes("/.ai/") ||
    normalized.startsWith(GOVERNANCE_DIR) ||
    normalized.endsWith("/state.yaml") ||
    normalized.endsWith("/gates.yaml")
  );
}

/**
 * Check if a path matches any of the protected path patterns.
 *
 * @param filePath - The file path to check
 * @returns `true` if the path is protected
 */
function isProtectedPath(filePath: string): boolean {
  const normalized = normalize(filePath).replace(/\\/g, "/");
  return PROTECTED_PATHS.some(pp => {
    const ppNorm = normalize(pp).replace(/\\/g, "/");
    return normalized.endsWith(ppNorm) || normalized === ppNorm;
  });
}

/**
 * Check if a bash command contains dangerous patterns.
 *
 * @param command - The command string to inspect
 * @returns `true` if the command is considered dangerous
 */
function isDangerousCommand(command: string): boolean {
  return DANGEROUS_COMMANDS.some(pattern => pattern.test(command));
}

/**
 * Safely read and parse a YAML file, returning null if not found or malformed.
 *
 * @param filePath - Absolute path to the YAML file
 * @returns Parsed object or null
 */
function safeReadYaml<T>(filePath: string): T | null {
  try {
    if (!existsSync(filePath)) return null;
    const raw = readFileSync(filePath, "utf-8");
    return parseDocument(raw).toJSON() as T;
  } catch {
    return null;
  }
}

// ── ContextController ────────────────────────────────────────────────────────

/**
 * Unified authorization engine implementing a 5-level priority decision chain.
 *
 * Every sensitive operation must call {@link authorize} before proceeding.
 * The chain is evaluated top-down; the first level that returns a non-null
 * result wins. If no level matches, the fail-closed default deny applies.
 *
 * Priority levels:
 * 1. High-risk operations → DENY
 * 2. Protected paths → ASK_USER
 * 3. Governance files check → ALLOW if no pending gate
 * 4. Task scope validation → check allowed_paths
 * 5. Default → fail-closed DENY
 */
export class ContextController {
  /**
   * Unified authorization entry point.
   *
   * Executes the 5-level priority decision chain and returns the first
   * non-null result. If all checks pass through, returns a default deny.
   *
   * @param request - The authorization request describing the intended action
   * @returns The authorization result with decision, reason, and check level
   */
  async authorize(request: AuthRequest): Promise<AuthResult> {
    // Level 1: High-risk operations
    const highRisk = await this._check_high_risk(request);
    if (highRisk) return highRisk;

    // Level 2: Protected paths
    const protected_ = await this._check_protected(request);
    if (protected_) return protected_;

    // Level 3: Governance files check
    const governance = await this._check_governance(request);
    if (governance) return governance;

    // Level 4: Task scope validation
    const taskScope = await this._check_task_scope(request);
    if (taskScope) return taskScope;

    // Level 5: Default deny
    return this._default_deny(request);
  }

  /**
   * Level 1 — High-risk operations check.
   *
   * Immediately denies operations that are inherently dangerous:
   * - INSTALL_PACKAGE and DELETE_FILE are always denied
   * - EXEC_BASH with dangerous commands (rm, format, etc.) is denied
   *
   * @param request - The authorization request
   * @returns AuthResult with DENY if high-risk, or null to pass to next level
   */
  private async _check_high_risk(request: AuthRequest): Promise<AuthResult | null> {
    // Always deny package installation
    if (request.action === Action.INSTALL_PACKAGE) {
      return {
        decision: Decision.DENY,
        reason: "High-risk: INSTALL_PACKAGE is not permitted. Use manual package management.",
        check_level: 1,
      };
    }

    // Always deny file deletion
    if (request.action === Action.DELETE_FILE) {
      return {
        decision: Decision.DENY,
        reason: "High-risk: DELETE_FILE is not permitted. Use manual file removal.",
        check_level: 1,
      };
    }

    // Check for dangerous bash commands
    if (request.action === Action.EXEC_BASH && request.command) {
      if (isDangerousCommand(request.command)) {
        return {
          decision: Decision.DENY,
          reason: `High-risk: EXEC_BASH command matches dangerous pattern: "${request.command}". Destructive system commands are blocked.`,
          check_level: 1,
        };
      }
    }

    return null;
  }

  /**
   * Level 2 — Protected paths check.
   *
   * Operations targeting protected files (AGENTS.md, .ai/state.yaml,
   * .ai/gates.yaml) require explicit user approval.
   *
   * MODIFY_GATE and MODIFY_STATE from non-system roles also require approval.
   *
   * @param request - The authorization request
   * @returns AuthResult with ASK_USER if protected, or null to pass to next level
   */
  private async _check_protected(request: AuthRequest): Promise<AuthResult | null> {
    // Check if target path is a protected file
    if (request.target_path && isProtectedPath(request.target_path)) {
      return {
        decision: Decision.ASK_USER,
        reason: `Protected path: "${request.target_path}" requires explicit user approval.`,
        check_level: 2,
      };
    }

    // MODIFY_GATE / MODIFY_STATE from non-system roles need approval
    if (request.action === Action.MODIFY_GATE || request.action === Action.MODIFY_STATE) {
      const systemRoles = ["system", "loop", "orchestrator"];
      const role = request.role_id?.toLowerCase() ?? "";
      if (!systemRoles.includes(role)) {
        return {
          decision: Decision.ASK_USER,
          reason: `Protected action: ${request.action} from non-system role "${request.role_id ?? "unknown"}" requires user approval.`,
          check_level: 2,
        };
      }
    }

    return null;
  }

  /**
   * Level 3 — Governance files check.
   *
   * Loads state.yaml and gates.yaml to determine governance status.
   * - If state.yaml not found → ALLOW (non-governance project)
   * - If no pending/blocked gates → ALLOW
   * - If target is a governance file (.ai/*) → ALLOW (Loop can write its own files)
   * - Otherwise → pass to next level
   *
   * @param request - The authorization request
   * @returns AuthResult with ALLOW if governance check passes, or null to continue
   */
  private async _check_governance(request: AuthRequest): Promise<AuthResult | null> {
    const root = resolve(request.project_root);

    // Try to load state — if not found, this is a non-governance project
    let state: ProjectState;
    try {
      state = await loadState(root);
    } catch {
      // No state.yaml → non-governance project, allow freely
      return {
        decision: Decision.ALLOW,
        reason: "No governance state found: project is not under Loop governance.",
        check_level: 3,
      };
    }

    // Load gates registry
    let gates: GatesRegistry;
    try {
      gates = await loadGates(root);
    } catch {
      // No gates.yaml → allow
      return {
        decision: Decision.ALLOW,
        reason: "No gates registry found: governance is not fully initialized.",
        check_level: 3,
      };
    }

    // Check for pending or blocked gates
    const pendingGates = gates.gates.filter(
      (g: GateDefinition) => g.status === "pending" || g.status === "blocked"
    );

    // Governance files are always writable by Loop itself
    if (request.target_path && isGovernancePath(request.target_path)) {
      return {
        decision: Decision.ALLOW,
        reason: "Governance file write: .ai/ files are always writable by Loop engine.",
        check_level: 3,
      };
    }

    // No pending gates → allow
    if (pendingGates.length === 0) {
      return {
        decision: Decision.ALLOW,
        reason: "No pending or blocked gates: all governance gates are passed.",
        check_level: 3,
      };
    }

    // Has pending gates but not a governance file → continue to task scope
    return null;
  }

  /**
   * Level 4 — Task scope validation.
   *
   * Loads the task graph and validates that:
   * - An active task exists
   * - The target path falls within the active task's allowed_paths
   *
   * @param request - The authorization request
   * @returns AuthResult with ALLOW/DENY based on task scope, or null if no task graph
   */
  private async _check_task_scope(request: AuthRequest): Promise<AuthResult | null> {
    const root = resolve(request.project_root);
    const taskGraphPath = join(root, TASK_GRAPH_FILE);

    // Load task graph
    const taskGraph = safeReadYaml<TaskGraph>(taskGraphPath);
    if (!taskGraph || !taskGraph.tasks || taskGraph.tasks.length === 0) {
      // No task graph → cannot validate scope, deny
      return {
        decision: Decision.DENY,
        reason: "No active task: task graph is missing or empty. Create a task before performing operations.",
        check_level: 4,
      };
    }

    // Find active task
    const activeTask = taskGraph.tasks.find(
      t => t.status === "active" || t.status === "in_progress"
    );

    if (!activeTask) {
      return {
        decision: Decision.DENY,
        reason: "No active task: all tasks are completed, pending, or blocked. Activate a task first.",
        check_level: 4,
      };
    }

    // If no target path specified, allow (e.g., read-only operations)
    if (!request.target_path) {
      return {
        decision: Decision.ALLOW,
        reason: `Task scope: active task "${activeTask.task_id}" — no target path to validate.`,
        allowed_paths: activeTask.allowed_paths,
        check_level: 4,
      };
    }

    // Check target path against allowed_paths
    const allowedPaths = activeTask.allowed_paths ?? [];

    if (allowedPaths.length === 0) {
      // No allowed_paths defined — deny by default for safety
      return {
        decision: Decision.DENY,
        reason: `Task scope: active task "${activeTask.task_id}" has no allowed_paths defined. Cannot verify target path.`,
        check_level: 4,
      };
    }

    // Normalize the target path for comparison
    const normalizedTarget = normalize(request.target_path).replace(/\\/g, "/");

    const inScope = allowedPaths.some(prefix => {
      const normalizedPrefix = normalize(prefix).replace(/\\/g, "/");
      return normalizedTarget.startsWith(normalizedPrefix);
    });

    if (!inScope) {
      return {
        decision: Decision.DENY,
        reason: `Task scope: target "${request.target_path}" is outside allowed paths [${allowedPaths.join(", ")}] for task "${activeTask.task_id}".`,
        allowed_paths: allowedPaths,
        check_level: 4,
      };
    }

    return {
      decision: Decision.ALLOW,
      reason: `Task scope: target "${request.target_path}" is within allowed paths for task "${activeTask.task_id}".`,
      allowed_paths: allowedPaths,
      check_level: 4,
    };
  }

  /**
   * Level 5 — Default deny (fail-closed).
   *
   * If no previous level produced a result, the operation is denied by default.
   * This is the fail-closed security posture: anything not explicitly allowed
   * is denied.
   *
   * @param request - The authorization request
   * @returns AuthResult with DENY
   */
  private _default_deny(request: AuthRequest): AuthResult {
    return {
      decision: Decision.DENY,
      reason: `Default deny: operation ${request.action} is not explicitly authorized.`,
      check_level: 5,
    };
  }
}

// ── Quick Auth Helper ────────────────────────────────────────────────────────

/**
 * Quick authorization check for common operations.
 *
 * Creates a ContextController and runs a single authorization check
 * with minimal parameters.
 *
 * @param root - The project root directory
 * @param action - The action to authorize
 * @param targetPath - Optional target file path
 * @returns The authorization result
 */
export async function quickAuth(
  root: string,
  action: Action,
  targetPath?: string,
): Promise<AuthResult> {
  const controller = new ContextController();
  return controller.authorize({
    action,
    target_path: targetPath,
    project_root: root,
  });
}

File: codex_loop/core/context_controller.py
"""
ContextController — Unified Authorization Engine for Loop Engine.

All sensitive operation authorization flows through this controller.
Hooks (gate_guard, loop_enforcement) should import and call this instead
of maintaining independent authorization logic.

Decision chain (priority from high to low):
1. High-risk operations → require independent gate → DENY
2. Protected paths (AGENTS.md, .zcode/config.json) → ASK_USER
3. Governance files (.ai/, .zcode/tools/, .zcode/skills/) →
   - No pending gate → ALLOW
   - Pending gate exists → check allowed_paths → ALLOW if in scope, else DENY
4. Task scope → match allowed_paths → ALLOW (if in scope), else DENY
5. Default → DENY (fail-closed)

Key: during a pending gate, writes targeting paths within the gate's
allowed_paths are ALLOWED. This solves the "design phase needs to write
design docs but is blocked by pending gate" deadlock.
"""
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from pathlib import Path


# ── Core Types ──────────────────────────────────────────────────────────

class Action(str, Enum):
    WRITE_FILE = "write_file"
    EXEC_BASH = "exec_bash"
    APPLY_PATCH = "apply_patch"
    MCP_TOOL_CALL = "mcp_tool_call"
    LAUNCH_ROLE = "launch_role"
    SUBMIT_EVIDENCE = "submit_evidence"
    STATE_TRANSITION = "state_transition"
    GATE_TRANSITION = "gate_transition"
    INSTALL = "install"
    UPGRADE = "upgrade"
    ROLLBACK = "rollback"


class Decision(str, Enum):
    ALLOW = "allow"
    DENY = "deny"
    ASK_USER = "ask_user"


@dataclass
class AuthRequest:
    action: Action
    target_path: str | None = None
    task_id: str | None = None
    role_id: str | None = None


@dataclass
class AuthResult:
    decision: Decision
    reason: str
    required_gate_id: str | None = None


# ── ContextController ───────────────────────────────────────────────────

class ContextController:
    """Unified authorization controller.

    All Hook calls route through authorize() for write/execute decisions.
    The decision chain runs checks in priority order; the first matching
    check produces the definitive result.
    """

    # Protected paths: writing to these always requires user confirmation.
    PROTECTED_PATHS: list[str] = ["AGENTS.md", ".zcode/config.json"]

    # Governance file prefixes: files under these are governance artifacts.
    GOVERNANCE_PREFIXES: list[str] = [".ai/", ".zcode/tools/", ".zcode/skills/"]

    # High-risk actions: require an independent gate (not self-approved).
    HIGH_RISK_ACTIONS: frozenset[Action] = frozenset({
        Action.INSTALL,
        Action.UPGRADE,
        Action.ROLLBACK,
        Action.GATE_TRANSITION,
    })

    # Decision-recording exemption: gates.yaml must always be writable
    # to avoid deadlock (user approves but AI cannot write the decision).
    DECISION_RECORDING_EXEMPT: frozenset[str] = frozenset({".ai/gates.yaml"})

    # Actions that are inherently safe without a file target.
    _NON_FILE_ACTIONS: frozenset[Action] = frozenset({
        Action.LAUNCH_ROLE,
        Action.SUBMIT_EVIDENCE,
        Action.STATE_TRANSITION,
    })

    def __init__(self, project_root: Path) -> None:
        self._project_root = Path(project_root).resolve()

    # ── Public API ────────────────────────────────────────────────────

    def authorize(self, request: AuthRequest) -> AuthResult:
        """Run the full decision chain and return an authorization result.

        Checks are executed in priority order; the first check that
        produces a result wins.
        """
        # 1. High-risk action check
        result = self._check_high_risk(request.action)
        if result is not None:
            return result

        # For actions without file targets that are not high-risk:
        # allow them (they are authorized at a higher level, e.g. executor).
        if request.target_path is None:
            if request.action in self._NON_FILE_ACTIONS:
                return AuthResult(
                    decision=Decision.ALLOW,
                    reason=f"Non-file action '{request.action.value}' allowed "
                           f"(authorized at host level)",
                )
            # Unrecognised action without target → deny
            return AuthResult(
                decision=Decision.DENY,
                reason=f"Action '{request.action.value}' without target_path "
                       f"cannot be authorized (fail-closed)",
            )

        # Decision-recording exemption: .ai/gates.yaml always allowed.
        rel = self._normalize_rel(request.target_path)
        if rel is not None and rel in self.DECISION_RECORDING_EXEMPT:
            return AuthResult(
                decision=Decision.ALLOW,
                reason="Decision-recording exemption: .ai/gates.yaml always writable",
            )

        # 2. Protected path check
        result = self._check_protected(request.target_path)
        if result is not None:
            return result

        # 3. Governance file check
        result = self._check_governance(request)
        if result is not None:
            return result

        # 4. Task scope check
        result = self._check_task_scope(request)
        if result is not None:
            return result

        # 5. Default deny (fail-closed)
        return AuthResult(
            decision=Decision.DENY,
            reason="No applicable authorization rule matched — deny by default (fail-closed)",
        )

    # ── Check 1: High-Risk Actions ────────────────────────────────────

    def _check_high_risk(self, action: Action) -> AuthResult | None:
        """High-risk actions require an independent gate approval.

        These actions cannot be self-approved by the AI and must go
        through an explicit gate with user or role approval.
        """
        if action in self.HIGH_RISK_ACTIONS:
            return AuthResult(
                decision=Decision.DENY,
                reason=(
                    f"High-risk action '{action.value}' requires an "
                    f"independent gate approval"
                ),
            )
        return None

    # ── Check 2: Protected Paths ──────────────────────────────────────

    def _check_protected(self, path: str | None) -> AuthResult | None:
        """Protected paths trigger ASK_USER to confirm the write.

        AGENTS.md and .zcode/config.json are critical project files
        that the user must explicitly approve modifications to.
        """
        if path is None:
            return None
        rel = self._normalize_rel(path)
        if rel is None:
            return None
        for protected in self.PROTECTED_PATHS:
            protected_norm = protected.replace("\\", "/")
            if rel == protected_norm:
                return AuthResult(
                    decision=Decision.ASK_USER,
                    reason=f"Protected path: '{rel}' requires user confirmation",
                )
        return None

    # ── Check 3: Governance Files ─────────────────────────────────────

    def _check_governance(self, request: AuthRequest) -> AuthResult | None:
        """Governance files (.ai/, .zcode/tools/, .zcode/skills/) logic:

        - No pending gate → ALLOW (governance maintenance is always OK).
        - Pending gate → check if target is in the gate's allowed_paths:
          * In allowed_paths → ALLOW (design docs can be written during review).
          * Not in allowed_paths → DENY (prevents bypassing the gate).
        """
        if request.target_path is None:
            return None
        rel = self._normalize_rel(request.target_path)
        if rel is None:
            return None
        if not self._is_governance_path(rel):
            return None

        pending = self._get_pending_gates()
        if not pending:
            return AuthResult(
                decision=Decision.ALLOW,
                reason="Governance file write, no pending gates",
            )

        # Pending gate(s) exist — check if target is in any gate's allowed_paths
        for gate in pending:
            gate_id = gate.get("id", "<unknown>")
            allowed = gate.get("allowed_paths", [])
            if isinstance(allowed, str):
                allowed = [allowed]
            for allowed_path in allowed:
                allowed_norm = self._normalize_scope_path(str(allowed_path))
                if self._path_matches(rel, allowed_norm):
                    return AuthResult(
                        decision=Decision.ALLOW,
                        reason=(
                            f"Governance file '{rel}' is in allowed_paths "
                            f"of pending gate '{gate_id}'"
                        ),
                    )

        # In allowed_paths of no pending gate → DENY
        return AuthResult(
            decision=Decision.DENY,
            reason=(
                f"Governance file '{rel}' is not in allowed_paths of any "
                f"pending gate: {[g.get('id') for g in pending]}"
            ),
        )

    # ── Check 4: Task Scope ───────────────────────────────────────────

    def _check_task_scope(self, request: AuthRequest) -> AuthResult | None:
        """Check if the target is within the active task's allowed_paths.

        - No task_id → DENY (cannot verify scope).
        - Task contract not found → DENY.
        - No allowed_paths in contract → DENY (fail-closed).
        - Target in allowed_paths → ALLOW.
        - Target not in allowed_paths → DENY.
        """
        if request.target_path is None:
            return None
        rel = self._normalize_rel(request.target_path)
        if rel is None:
            return None

        task_id = request.task_id or self._get_current_task_id()
        if task_id is None:
            return AuthResult(
                decision=Decision.DENY,
                reason="No task_id available — cannot verify write scope",
            )

        contract = self._load_task_contract(task_id)
        if contract is None:
            return AuthResult(
                decision=Decision.DENY,
                reason=f"Task contract for '{task_id}' not found",
            )

        allowed = contract.get("allowed_paths", [])
        if not allowed:
            return AuthResult(
                decision=Decision.DENY,
                reason=f"Task '{task_id}' has no allowed_paths defined",
            )

        for allowed_path in allowed:
            allowed_norm = self._normalize_scope_path(str(allowed_path))
            if self._path_matches(rel, allowed_norm):
                return AuthResult(
                    decision=Decision.ALLOW,
                    reason=f"Path '{rel}' is in allowed scope of task '{task_id}'",
                )

        return AuthResult(
            decision=Decision.DENY,
            reason=(
                f"Path '{rel}' is not in allowed_paths of task '{task_id}': "
                f"{allowed}"
            ),
        )

    # ── Helpers ───────────────────────────────────────────────────────

    def _normalize_rel(self, target: str) -> str | None:
        """Convert target path to a posix-style relative path from project root.

        Returns None if the target cannot be resolved or is outside the
        project root.
        """
        try:
            target_path = Path(target)
            if not target_path.is_absolute():
                target_path = self._project_root / target_path
            rel = target_path.resolve().relative_to(self._project_root.resolve())
            return rel.as_posix()
        except (ValueError, OSError):
            return None

    def _is_governance_path(self, rel: str) -> bool:
        """Check whether a relative path is a governance artifact."""
        rel = rel.replace("\\", "/")
        for prefix in self.GOVERNANCE_PREFIXES:
            prefix_norm = prefix.replace("\\", "/")
            if rel.startswith(prefix_norm):
                return True
        return False

    @staticmethod
    def _path_matches(rel: str, scope: str) -> bool:
        """Check if rel is equal to or inside scope (directory prefix match)."""
        scope = scope.rstrip("/")
        if rel == scope:
            return True
        if rel.startswith(scope + "/"):
            return True
        return False

    @staticmethod
    def _normalize_scope_path(raw: str) -> str:
        """Normalize a scope/allowed path: backslashes to forward, strip ./ prefix.

        Uses startswith-based check (not lstrip) because lstrip treats its
        argument as a character set, which strips all leading '.' and '/'
        characters rather than the './' prefix.
        """
        norm = raw.replace("\\", "/")
        if norm.startswith("./"):
            norm = norm[2:]
        return norm

    # ── File I/O ──────────────────────────────────────────────────────

    def _load_state(self) -> dict:
        """Load .ai/state.yaml; return empty dict on any failure."""
        state_path = self._project_root / ".ai" / "state.yaml"
        if not state_path.exists():
            return {}
        try:
            data = self._yaml_load(state_path)
            return data if isinstance(data, dict) else {}
        except Exception:
            return {}

    def _load_gates(self) -> list[dict]:
        """Load the gates list from .ai/gates.yaml.

        Returns an empty list if the file does not exist or cannot be parsed.
        """
        gates_path = self._project_root / ".ai" / "gates.yaml"
        if not gates_path.exists():
            return []
        try:
            data = self._yaml_load(gates_path) or {}
            gates = data.get("gates", [])
            return gates if isinstance(gates, list) else []
        except Exception:
            return []

    def _load_task_contract(self, task_id: str) -> dict | None:
        """Load a task contract from .ai/tasks/{task_id}.md.

        Parses YAML-like front-matter fields:
        - allowed_paths: list of path strings
        - developer_agent_id / reviewer_agent_id: agent IDs
        """
        task_path = self._project_root / ".ai" / "tasks" / f"{task_id}.md"
        if not task_path.exists():
            return None

        text = task_path.read_text(encoding="utf-8")
        contract: dict = {
            "allowed_paths": [],
            "developer_agent_id": None,
            "reviewer_agent_id": None,
        }

        in_allowed_section = False
        for line in text.splitlines():
            stripped = line.strip()
            if stripped.startswith("allowed_paths:") or stripped.startswith("allowed_actions:"):
                in_allowed_section = True
                continue
            if in_allowed_section and stripped.startswith("- "):
                path = stripped[2:].strip().strip('"')
                contract["allowed_paths"].append(path)
            elif in_allowed_section and not stripped.startswith("- "):
                in_allowed_section = False
            if stripped.startswith("developer_agent_id:"):
                contract["developer_agent_id"] = stripped.split(":", 1)[1].strip().strip('"')
            elif stripped.startswith("reviewer_agent_id:"):
                contract["reviewer_agent_id"] = stripped.split(":", 1)[1].strip().strip('"')

        return contract

    def _get_current_task_id(self) -> str | None:
        """Read current_task_id from .ai/state.yaml."""
        state = self._load_state()
        tid = state.get("current_task_id")
        return str(tid) if tid else None

    def _get_pending_gates(self) -> list[dict]:
        """Return gates with status == 'pending'."""
        gates = self._load_gates()
        return [g for g in gates if isinstance(g, dict) and g.get("status") == "pending"]

    @staticmethod
    def _yaml_load(file_path: Path) -> dict | list | None:
        """Load a YAML file, preferring PyYAML with a text-scan fallback.

        The fallback handles the common case of flat key-value pairs and
        simple list-of-dicts structures used in Loop governance files.
        """
        text = file_path.read_text(encoding="utf-8")
        try:
            import yaml  # type: ignore
            return yaml.safe_load(text)
        except ImportError:
            pass
        return ContextController._naive_yaml_parse(text)

    @staticmethod
    def _naive_yaml_parse(text: str) -> dict:
        """Conservative text-based YAML parser for Loop governance files.

        Handles:
        - Top-level scalar keys
        - Nested lists of dicts under keys (gates, tasks)
        - Within dicts: scalar keys and list values (allowed_paths)

        Uses the original line (not stripped) for indent-sensitive matches.
        Only strips when checking plain content (starts-with, key detection).
        """
        import re

        result: dict = {}
        lines = text.splitlines()

        # Pass 1: top-level scalars (no indentation required)
        for line in lines:
            m = re.match(r"^([A-Za-z_][A-Za-z0-9_]*):\s*(.*)$", line)
            if m:
                key = m.group(1)
                value = m.group(2).strip().strip('"').strip("'")
                if value in ("null", "~", ""):
                    result[key] = None
                else:
                    result[key] = value

        # Pass 2: list-of-dicts (gates, tasks)
        list_keys = ["gates", "tasks"]
        for list_key in list_keys:
            items: list[dict] = []
            current_item: dict | None = None
            in_list = False
            in_allowed = False

            for line in lines:
                stripped = line.strip()
                # Detect list key start (no indentation for the key itself)
                if re.match(rf"^{list_key}:\s*$", stripped):
                    in_list = True
                    current_item = None
                    continue
                if in_list:
                    # New item: "- id:" or "- name:" etc.
                    # Use original line so leading whitespace is preserved for the regex.
                    m_item = re.match(r"^\s*-\s+(\S+):\s*(.*)$", line)
                    if m_item:
                        if current_item is not None:
                            items.append(current_item)
                        current_item = {}
                        k = m_item.group(1)
                        v = m_item.group(2).strip().strip('"').strip("'")
                        current_item[k] = v if v not in ("", "~", "null") else None
                        in_allowed = False
                        continue
                    if current_item is not None:
                        # Scalar field inside item: must be indented ≥2 spaces.
                        # Use original line so leading whitespace is preserved.
                        m_field = re.match(r"^\s{2,}(\S+):\s*(.*)$", line)
                        if m_field:
                            k = m_field.group(1)
                            v = m_field.group(2).strip().strip('"').strip("'")
                            if k == "allowed_paths":
                                in_allowed = True
                                current_item[k] = []
                            elif v in ("", "~", "null"):
                                current_item[k] = None
                            else:
                                current_item[k] = v
                            continue
                        # List item under a field (e.g. allowed_paths): ≥4 spaces indent.
                        # Use original line so leading whitespace is preserved.
                        m_list_item = re.match(r"^\s{4,}-\s+(.+)$", line)
                        if m_list_item and in_allowed:
                            current_item.setdefault("allowed_paths", []).append(
                                m_list_item.group(1).strip().strip('"').strip("'")
                            )
                            continue
                        # Non-matching line may end current list membership
                        if stripped and not stripped.startswith("#"):
                            # Could be start of a different top-level section
                            if re.match(r"^[A-Za-z_][A-Za-z0-9_]*:\s*$", stripped):
                                in_list = False
                                if current_item is not None:
                                    items.append(current_item)
                                current_item = None
                                continue

            if current_item is not None:
                items.append(current_item)
            if items:
                result[list_key] = items

        return result
